packages/@uppy/companion-client/src/Provider.ts

```diff
@@ -0,0 +1,445 @@
+} from './RequestClient.ts'
+import * as tokenStorage from './tokenStorage.ts'
+
+// TODO: remove deprecated options in next major release
+export type Opts = {
+  /** @deprecated */
+  serverUrl?: string
+  /** @deprecated */
+  serverPattern?: string
+  companionUrl: string
+  companionAllowedHosts?: string | RegExp | Array<string | RegExp>
+  storage?: typeof tokenStorage
+  pluginId: string
+  name?: string
+  supportsRefreshToken?: boolean
+  provider: string
+}
+
+class ProviderPlugin<M extends Meta, B extends Body> extends BasePlugin<
+  Opts,
+  M,
+  B
+> {
+  files: UppyFile<M, B>[]
+
+  storage: typeof tokenStorage
+}
+
+const getName = (id: string) => {
+  return id
+    .split('-')
+    .map((s) => s.charAt(0).toUpperCase() + s.slice(1))
+    .join(' ')
+}
+
+function getOrigin() {
+  // eslint-disable-next-line no-restricted-globals
+  return location.origin
+}
+
+function getRegex(value: string | RegExp) {
+  if (typeof value === 'string') {
+    return new RegExp(`^${value}$`)
+  }
+  if (value instanceof RegExp) {
+    return value
+  }
+  return undefined
+}
+
+function isOriginAllowed(
+  origin: string,
+  allowedOrigin: string | Array<string | RegExp>,
+) {
+  const patterns = Array.isArray(allowedOrigin)
+    ? allowedOrigin.map(getRegex)
+    : [getRegex(allowedOrigin)]
+  return patterns.some(
+    (pattern) => pattern?.test(origin) || pattern?.test(`${origin}/`),
+  ) // allowing for trailing '/'
+}
+
+export default class Provider<
+  M extends Meta,
+  B extends Body,
+> extends RequestClient<M, B> {
+  #refreshingTokenPromise: Promise<void> | undefined
+
+  provider: string
+
+  id: string
+
+  name: string
+
+  pluginId: string
+
+  tokenKey: string
+
+  companionKeysParams?: Record<string, string>
+
+  preAuthToken: string | null
+
+  supportsRefreshToken: boolean
+
+  constructor(uppy: Uppy<M, B>, opts: Opts) {
+    super(uppy, opts)
+    this.provider = opts.provider
+    this.id = this.provider
+    this.name = this.opts.name || getName(this.id)
+    this.pluginId = this.opts.pluginId
+    this.tokenKey = `companion-${this.pluginId}-auth-token`
+    this.companionKeysParams = this.opts.companionKeysParams
+    this.preAuthToken = null
+    this.supportsRefreshToken = opts.supportsRefreshToken ?? true // todo false in next major
+  }
+
+  async headers(): Promise<Record<string, string>> {
+    const [headers, token] = await Promise.all([
+      super.headers(),
+      this.#getAuthToken(),
+    ])
+    const authHeaders: Record<string, string> = {}
+    if (token) {
+      authHeaders['uppy-auth-token'] = token
+    }
+
+    if (this.companionKeysParams) {
+      authHeaders['uppy-credentials-params'] = btoa(
+        JSON.stringify({ params: this.companionKeysParams }),
+      )
+    }
+    return { ...headers, ...authHeaders }
+  }
+
+  onReceiveResponse(response: Response): Response {
+    super.onReceiveResponse(response)
+    const plugin = this.uppy.getPlugin(this.pluginId)
+    if (!plugin) return response
```

Comment:
That was added in this PR, could we instead add non-null assertions and we can take care of it in the `4.x` branch
```suggestion
```